### DISCUSSION TITLE
[mini.comment] Keep visual selection after toggling comment

### DISCUSSION THREAD

**Discussion by louwers** [nvim-mini/mini.nvim#1313] [Q&A]
I would like to keep the visual selection after commenting (out) a block of code.

Also I would not like to change lines.

Is this possible?

**Comment by echasnovski** [comment]
I am afraid I understand only part of the question.

> I would like to keep the visual selection after commenting (out) a block of code.

So you'd like to have it in Visual mode like this:
- If text is made comment, then preserve visual selection.
- If text is made not comment, then not preserve visual selection.

Having this behavior *is* possible, but will not be officially supported. Here is a snippet:

```lua
local is_vis_mode = false
local maybe_restore_vis_mode = function(context)
  if is_vis_mode and context.action == 'comment' then vim.schedule(function() vim.cmd('normal! gv') end) end
  is_vis_mode = false
end
local vis_operator_rhs = function()
  is_vis_mode = true
  return MiniComment.operator()
end

require('mini.comment').setup({ hooks = { post = maybe_restore_vis_mode } })
vim.keymap.set('x', 'gc', vis_operator_rhs, { desc = 'Comment selection', expr = true })
```

This basically overcomes the issue of determining whether the action meets the criteria for preserving visual selection (via `normal! gv`) by using custom mapping in Visual mode (via explicit `MiniComment.operator()` call which I'd personally not recommend).

------

> Also I would not like to change lines.

I don't really understand this. Could you elaborate?

**Reply by louwers** [reply]
Basically, I want the same behavior as other editors such as VS Code.

**Reply by louwers** [reply]
https://github.com/user-attachments/assets/59e272c4-3171-417c-ac46-26544a551ef2

**Reply by echasnovski** [reply]
> Basically, I want the same behavior as other editors such as VS Code.

This is Neovim which itself has a long history (including Vim) of how it handles commenting and 'mini.comment' represents exactly that.

As far as I can tell with the brief VS Code tests, 'mini.comment' behaves more like `<C-/>` (i.e. per-line toggle comment which stops selection) and not like `<C-S-a>` (i.e. toggle block comment which preserves selection).

The closest to VS Code behavior of `<C-S-a>` can be achieved with the solution above, but probably also `and context.action == 'comment'` condition needs to be removed (to preserve selection regardless of the action performed). This will work only on per line basis as ['mini.comment' doesn't support block comments](https://github.com/echasnovski/mini.nvim/blob/919b953f044d97a2243cc2d1f0b316a199434127/doc/mini-comment.txt#L21).

My sincere suggestion would be to get used to the mini.comment' behavior as most Neovim (and Vim) operators don't preserve visual selection after operating on it.

**Reply by louwers** [reply]
Thanks for your detailed answers!
